### ISSUE TITLE
ADD and COPY commands short-circuts the build

### ISSUE THREAD

**Issue by torbjornvatn** [JoTrdl/grunt-dock#1]
I'm trying to build a docker image with grunt-dock but whenever I get to a ADD or COPY step the build is aborted and `>> Build successfuly done` is printed.

This is an example Dockerfile

``` dockerfile
FROM dockerfile/nodejs
MAINTAINER Torbjørn Vatn <torbjorn.vatn@penger.no>

RUN npm install -g grunt-cli

ADD package.json /data/package.json

RUN ls
```

**Comment by jmweir** [comment]
I've confirmed this issue as well.

**Comment by jmweir** [comment]
Taking a look at process() function in lib/build.js reveals that 'dockerfile' parameter may either be a path to a tar file (that would contain your additional files) or path to Dockerfile only, in which case it is converted to a tar that contains only the Dockerfile. I wonder if there's a good workflow to have grunt generate a tar of your directory, though it seems you'd have to write an intermediate tar file to disk.

This modified process() function allows for 'dockerfile' parameter to be a directory path. This relies on the NPM 'tar-fs' module also being pulled in.

```
 var process = function(tag, image, callback) {

      // Create the options object for the build
      var imageOptions = image.options || {};
      var buildOpts = utils.merge({t: tag}, imageOptions.build || {});

      if (fs.lstatSync(image.dockerfile).isDirectory()) {

          tarfs.pack(image.dockerfile).pipe(zlib.createGzip()).pipe(concat(function (data) {
              build(data, buildOpts, callback);
          }));

      } else {
          var content;
          try {
              content = fs.readFileSync(image.dockerfile);
          } catch (e) {
              return done(e);
          }

          var fileExt = path.extname(image.dockerfile);
          if (fileExt) {
              build(content, buildOpts, callback);
          }
          else {
              var pack = tar.pack();
              pack.entry({name: 'Dockerfile'}, content);
              pack.finalize();

              pack.pipe(zlib.createGzip()).pipe(concat(function (data) {
                  build(data, buildOpts, callback);
              }));
          }
      }
  };
```

**Comment by jmweir** [comment]
Revised the function so that it packs the parent directory of the Dockerfile. You continue to specify the Dockerfile path as normally.

```
  var process = function(tag, image, callback) {

      // Create the options object for the build
      var imageOptions = image.options || {};
      var buildOpts = utils.merge({t: tag}, imageOptions.build || {});

      var fileExt = path.extname(image.dockerfile);

      if (fileExt) {
          var content;

          try {
              content = fs.readFileSync(image.dockerfile);
          } catch (e) {
              return done(e);
          }

          build(content, buildOpts, callback);
      } else {
          tarfs.pack(path.dirname(path.resolve(image.dockerfile))).pipe(zlib.createGzip()).pipe(concat(function (data) {
              build(data, buildOpts, callback);
          }));
      }
  };
```

**Comment by JoTrdl** [comment]
Thanks for you feedback. I will prepare a new version soon.

**Comment by torbjornvatn** [comment]
@JoTrdl sweet!

**Comment by jmweir** [comment]
Sounds great. I posted my approach as a pull request in case that helps.

**Comment by JoTrdl** [comment]
Just published the new version (1.0.2) on npm.
Many thanks!
